lms/services/blackboard_api/client.py

```diff
@@ -88,10 +78,74 @@ def list_files(self, course_id, folder_id=None):
                     "size": file["size"],
                     "parent_lms_id": file["parentId"],
                 }
-                for file in results
+                for file in files
             ]
         )
 
+    def list_files(self, course_id, folder_id=None):
+        """Return the list of files in the given course or folder."""
+        results = []
+
+        path = self._list_files_url(course_id, folder_id)
+
+        for _ in range(PAGINATION_MAX_REQUESTS):
+            response = self._api.request("GET", path)
+            results.extend(BlackboardListFilesSchema(response).parse())
+            path = response.json().get("paging", {}).get("nextPage")
+            if not path:
+                break
+
+        self._store_files(course_id, results)
+
+        return results
+
+    def _get_all_folders_files(self, course_id, results):
+        # Get all the folder URL from `results`
+        folder_urls = [
+            self._api._api_url(self._list_files_url(course_id, result["id"]))
+            for result in results
+            if result["type"] == "Folder"
+        ]
+
+        # Query all folders at once
+        responses = self._request.find_service(name="async_oauth_http").request(
+            "GET", folder_urls
+        )
+
+        for response in responses:
+            # Get the first page result
+            folder_results = BlackboardListFilesSchema(response).parse()
+            # Add any extra pages
+            for _ in range(PAGINATION_MAX_REQUESTS - 1):
+                path = response.sync_json.get("paging", {}).get("nextPage")
+                if not path:
+                    break
+                response = self._api.request("GET", path)
+                folder_results.extend(BlackboardListFilesSchema(response).parse())
+
+            # For each folder, get it children too
+            results.extend(self._get_all_folders_files(course_id, folder_results))
+
+        return results
+
+    def list_all_files(self, course_id):
+        results = []
+
+        # Get this level all pages
+        path = self._list_files_url(course_id)
+        for _ in range(PAGINATION_MAX_REQUESTS):
+            response = self._api.request("GET", path)
+            results.extend(BlackboardListFilesSchema(response).parse())
+            path = response.json().get("paging", {}).get("nextPage")
+            if not path:
+                break
+
+        results += self._get_all_folders_files(course_id, results)
```

Comment:
Disregard this code in particular but I reckon our options are:

- Naively get all pages and children sequentially
- Get all pages sequentially but all children at once with async
- Get all children at once and all the "next pages" we can for each of them using async.